### PR TITLE
[Feature] Support invoking score infer and emb_sim using API.

### LINKED ISSUE #151: [Bug] In PyTorch 1.6, Integer division using div is no longer supported
In PyTorch 1.6, Integer division using div is no longer supported, and in a future release div will perform true division as in Python 3. Use torch.true_divide() or torch.floor_divide() (// in Python), instead.

Mentioned in https://discuss.dgl.ai/t/knowledge-graphs-dglke/1332.

### PR DESCRIPTION
**Pull request by classicsong**
This PR introduce a general framework named BasicGEModel (Basic graph embedding model) following the discussion in #82 . Graph Embedding models, either knowledge graph embedding models or GNN based graph embedding models will follow this framework.

This PR focus on providing Python API support for model inference. To do edge score inference, one can follow the following operations:
```
gamma = 2.0
transe_model = TransEModel('cpu', gamma)
transe_model.load(PATH_TO_MODEL)
transe_model.attach_graph(PATH_TO_GRAPH)
result = transe_model.link_predict(...)
```
To do embedding similarity inference,  one can follow the following operations:
```
gamma = 2.0
transe_model = TransEModel('cpu', gamma)
transe_model.load(PATH_TO_MODEL)
# we omit attach_graph, as the graph structure is not used
result = transe_model.embed_sim(...)
```

Further, in the link_predict inference, we will support different edge excluding mode:

  * None, existing edges are not excluded.
  * mask, existing edges are masked using a th.bool tensor.
  * exclude, existing edges are excluded in the result.

This PR also fix #151

### CODE REVIEW

File: python/dglke/models/ke_model.py [hunk 1]
```diff
@@ -0,0 +1,913 @@
...
+        head_emb = self.entity_embed[head]
+        rel_emb = self.relation_embed[rel]
+        tail_emb = self.entity_embed[tail]
+
+        num_head = head.shape[0]
+        num_rel = rel.shape[0]
+        num_tail = tail.shape[0]
+
+        score = []
+        if triplet_wise:
```
Comment (zheng-da) [inline python/dglke/models/ke_model.py:137]: i think you need to write comments to explain this. I'm not sure what this option means.

File: python/dglke/models/ke_model.py [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+
+                @property
+                def dst(self):
+                    return self._tobj
+
+                @property
+                def data(self):
+                    return self._robj
+
+            for i in range((num_head + batch_size - 1) // batch_size):
```
Comment (zheng-da) [inline python/dglke/models/ke_model.py:159]: what does this iterate over?

File: python/dglke/models/pytorch/ke_tensor.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+from _thread import start_new_thread
+import traceback
+from functools import wraps
+
+from .. import *
+
+from .tensor_models import thread_wrapped_func
+
+@thread_wrapped_func
+def async_update(num_thread, emb, queue):
```
Comment (zheng-da) [inline python/dglke/models/pytorch/ke_tensor.py:42]: i don't understand why the code is moved here?

File: python/dglke/models/pytorch/tensor_models.py [hunk 4]
```diff
@@ -129,7 +132,7 @@ def async_update(args, emb, queue):
     """Asynchronous embedding update for entity embeddings.
```
Comment (zheng-da) [inline python/dglke/models/pytorch/tensor_models.py:132]: async_update was defined here.

File: python/dglke/models/pytorch/ke_tensor.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+            if gpu_id >= 0:
+                std = std.cuda(gpu_id)
+            std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
+            tmp = (-clr * grad_values / std_values)
+            if tmp.device != device:
+                tmp = tmp.to(device)
+            emb.emb.index_add_(0, grad_indices, tmp)
+
+
+class KGEmbedding:
```
Comment (zheng-da) [inline python/dglke/models/pytorch/ke_tensor.py:83]: where is it used for training?

File: python/dglke/models/pytorch/ke_tensor.py (hunk 3 above)
Comment (classicsong) [reply python/dglke/models/pytorch/ke_tensor.py:42]: I just copy paste to here. Maybe we can move them when doing training code refactor.

File: python/dglke/models/pytorch/ke_tensor.py (hunk 5 above)
Comment (classicsong) [reply python/dglke/models/pytorch/ke_tensor.py:83]: Currently only the load() method is used to load the embeddings.

File: python/dglke/models/ke_model.py [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+                                                   if (i + 1) * batch_size < num_head \
+                                                   else num_head]
+                sr_emb = rel_emb[i * batch_size : (i + 1) * batch_size \
+                                                  if (i + 1) * batch_size < num_head \
+                                                  else num_head]
+                st_emb = tail_emb[i * batch_size : (i + 1) * batch_size \
+                                                   if (i + 1) * batch_size < num_head \
+                                                   else num_head]
+                edata = FakeEdge(sh_emb, sr_emb, st_emb, self._device)
+                score.append(self._score_func.edge_func(edata)['score'].to(th.device('cpu')))
```
Comment (zheng-da) [inline python/dglke/models/ke_model.py:173]: is it necessary to do it in a mini-batch fashion?

File: python/dglke/models/ke_model.py [hunk 7]
```diff
@@ -0,0 +1,976 @@
...
+        res_head = []
+        res_rel = []
+        res_tail = []
+        res_score = []
+        result = []
+        if exclude_mode == 'exclude':
+            # exclude existing edges
+            cur_k = 0
+            batch_size = topk
+            while (cur_k < num_triples):
```
Comment (zheng-da) [inline python/dglke/models/ke_model.py:219]: is this really fast?

File: python/dglke/models/ke_model.py [hunk 8]
```diff
@@ -0,0 +1,976 @@
...
+        """ Generate topk most relevent triplets and corresponding scores.
+
+            It takes following steps:
+
+              1) find topk elements
+              2) sort topk elements in descending order
+              3) call _exclude_pos if figure out existing edges
+        """
+        if exclude_mode == 'exclude':
+            if idx.shape[0] < topk * 4: # TODO(xiangsx): Find a better value of topk * n
```
Comment (zheng-da) [inline python/dglke/models/ke_model.py:402]: it finds 4k edges and removes positive edges? this seems a lot of extra work.

Comment (zheng-da) [inline python/dglke/models/ke_model.py:402]: do we need to check this? can we always use the else branch?

File: python/dglke/models/ke_model.py (hunk 7 above)
Comment (classicsong) [reply python/dglke/models/ke_model.py:219]: Not that fast.

File: python/dglke/models/ke_model.py (hunk 8 above)
Comment (classicsong) [reply python/dglke/models/ke_model.py:402]: You cannot do topk when k is larger than len(score)
